### PR TITLE
Added integration tests for src/graphql/types/Mutation/updateTag

### PR DESCRIPTION
**Pull request by MonishPuttu**
Fixes: #3847

## Test Coverage

**15 Test Cases Written:**

### Authentication (2 tests)
- Unauthenticated user returns error
- Deleted user with valid token returns error

### Input Validation (2 tests)
- Invalid UUID format for tag ID
- Invalid UUID format for folder ID

### Resource Validation (2 tests)
- Tag not found error
- Folder tag not found error

### Business Logic (2 tests)
- Cross-organization folder assignment prevention
- Duplicate tag name within same organization prevention

### Authorization (2 tests)
- Unauthorized non-member attempting update
- Unauthorized regular member attempting update (requires admin role)

### Error Handling (1 test)
- Database update operation failure handling

### Success Scenarios (2 tests)
- Successfully update tag name only
- Successfully update tag with valid folderId from same organization

### Edge Cases (2 tests)
- Successfully handle tag names with special characters
- Successfully handle unicode characters in tag name

## Coverage Achieved

- **Statement:** 100% 
- **Branch:** 100% 
- **Function:** 100% 
- **Lines:** 100% 

## Files Modified

- src/graphql/types/Mutation/updateTag.ts - Fixed folderId validation bug
- test/graphql/types/Mutation/updateTag.test.ts - Complete test suite (15 tests)

## Summary by CodeRabbit

* **Bug Fixes**
  * Corrected folder validation during tag updates so existing folders are correctly recognized.

* **Tests**
  * Added comprehensive tests for tag updates covering authentication, argument validation, resource existence, authorization levels, simulated database failures, successful updates (including removing folder assignment), and special/unicode name cases.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by MonishPuttu** [comment]
@coderabbitai review

**Comment by palisadoes** [comment]
@coderabbitai full review

Perform a comprehensive analysis and reevaluate all resolved items with the following criteria:

**Test Coverage Analysis:**
1. For each modified or new test file, verify:
   - All edge cases and use cases are covered (success paths, error paths, boundary conditions)
   - Test assertions are correct and meaningful
   - 100% code coverage is achieved for the associated implementation files
   - No redundant or duplicate test cases exist
   
2. For each modified or new source file, identify:
   - Specific lines of code lacking test coverage
   - Functions, branches, or conditions that need additional test cases
   - Mock/stub requirements that may be missing

**Code Quality Assessment:**
3. Identify any files that should be excluded:
   - Accidentally committed files (IDE configs, env files, temporary files)
   - Files unrelated to the PR scope
   - Generated files that should be in .gitignore

**Requirement Validation:**
4. Verify the PR fully addresses the linked issue:
   - All acceptance criteria from the issue are met
   - Implementation matches the described solution approach
   - No scope creep or unrelated changes are included

5. Confirm compliance with all previous review feedback:
   - All resolved comments have been properly addressed
   - No regressions of previously suggested improvements
   - Coding standards and conventions are followed

**Review Decision:**
- Request changes (CHANGES_REQUESTED) if ANY of the following apply:
  - Missing test coverage or invalid tests
  - Code quality issues, security concerns, or bugs
  - Incomplete implementation of issue requirements
  - Non-compliance with previous feedback
  - Presence of unnecessary files

- Approve (APPROVED) only if:
  - All criteria above are fully satisfied
  - No changes, even trivial ones, are required

**Comment by palisadoes** [comment]
@coderabbitai full review using this prompt:
- https://github.com/PalisadoesFoundation/talawa-api/pull/4055#issuecomment-3684158686

**Comment by MonishPuttu** [comment]
@coderabbitai, I have some questions about the feedback that I'd like your help clarifying:

1. There seems to be two different Mutations.
Could you verify if there are two separate mutations in the codebase?

1. src/graphql/types/Mutation/updateTag.ts (this PR)
2. src/graphql/types/Mutation/updateTagFolder.ts (different file)

Question: Are the suggested folderId tests actually meant for updateTagFolder instead of updateTag?

2. Database Schema Mismatch?

Looking at the implementation at line 99-108:
const existingTagFolder = await ctx.drizzleClient.query.tagsTable.findFirst({
And the database schema at src/drizzle/tables/tags.ts:24-27:
folderId: uuid("folder_id").references(() => tagFoldersTable.id, {

Questions:
Should the validation query tagFoldersTable instead of tagsTable?
Would this cause a foreign key constraint violation in the database even if validation passes?

3. Coverage Calculation Question

The coverage report shows:
updateTag.ts: 100% (all metrics)

Questions:
Do the error path tests (non-existent folder, cross-org folder) already execute lines 91-124?
Would a success path test add any new code execution coverage, or just test database operations?
Does coverage measure code execution or database transaction success?

4. Test Implementation Clarification

Your suggested test uses:
const createFolderResult = await mercuriusClient.mutate(Mutation_createTag, {
Questions:
Does Mutation_createTag create a tag in tagsTable or a folder in tagFoldersTable?
Should this be Mutation_createTagFolder instead?

5. What Am I Missing?

Final questions:
Given the 100% coverage report, which specific lines are you seeing as uncovered?
Is there a difference between branch coverage and line coverage that I'm not understanding?
Should I be testing database FK constraint enforcement, or just the application logic?
I want to ensure I'm implementing the correct tests. Could you help clarify these points?

**Comment by MonishPuttu** [comment]
@coderabbitai, Your acknowledgment of the implementation bug validates the testing approach taken in this PR.
​
Response to Your Questions:

1. Should you create an issue for the folderId validation bug?
   Yes, That would be helpful for tracking the implementation fix. The issue should note:
- Bug location: src/graphql/types/Mutation/updateTag.ts
- Fix: Change query.tagsTable.findFirst → query.tagFoldersTable.findFirst

2. Should you approve this PR as-is?
    Yes, please approve.

**Comment by palisadoes** [comment]
@coderabbitai

The aim of testing is to validate the operation of the file under test and modify it if errors are found. How can this prompt be updated to ensure that any errors in the file being tested must be fixed in the PR?

- https://github.com/PalisadoesFoundation/talawa-api/pull/4055#issuecomment-3684158686

**Comment by palisadoes** [comment]
@coderabbitai full review

Perform a comprehensive analysis and reevaluate all resolved items with the following criteria:

**Test Coverage Analysis:**
1. For each modified or new test file, verify:
   - All edge cases and use cases are covered (success paths, error paths, boundary conditions)
   - Test assertions are correct and meaningful
   - 100% code coverage is achieved for the associated implementation files
   - No redundant or duplicate test cases exist
   
2. For each modified or new source file, identify:
   - Specific lines of code lacking test coverage
   - Functions, branches, or conditions that need additional test cases
   - Mock/stub requirements that may be missing

**Code Quality Assessment:**
3. Identify any files that should be excluded:
   - Accidentally committed files (IDE configs, env files, temporary files)
   - Files unrelated to the PR scope
   - Generated files that should be in .gitignore

**Implementation Correctness:** 
4. For the specific implementation file(s) being tested:
   - Verify the implementation logic is correct and free of bugs
   - Validate that queries target the correct database tables/entities
   - Check that foreign key relationships are correctly enforced
   - Confirm business logic constraints are properly implemented
   - Any bugs, logic errors, or incorrect implementations discovered in the file(s) under test MUST be fixed in this PR before approval

**Requirement Validation:**
5. Verify the PR fully addresses the linked issue:
   - All acceptance criteria from the issue are met
   - Implementation matches the described solution approach
   - No scope creep or unrelated changes are included

6. Confirm compliance with all previous review feedback:
   - All resolved comments have been properly addressed
   - No regressions of previously suggested improvements
   - Coding standards and conventions are followed

7. **Review Decision:**
    1. When critical issues are found, clearly state that changes are required in your response but DO NOT append  the APPROVED tag.
    1.  Only append  the APPROVED tag when ALL criteria are satisfied and zero changes are required:
        1. Complete test coverage (≥95%)
        1. No code quality issues, security concerns, or bugs
        1. Full implementation of issue requirements
        1. No bugs or logic errors in the implementation file(s) under test
        1. Compliance with all previous feedback
        1. No unnecessary files

8. Approve (APPROVED) only if:
  - All criteria above are fully satisfied
  - The implementation file(s) under test are correct and bug-free [NEW]
  - No changes, even trivial ones, are required

**Note:** Test-focused PRs should maintain clear scope boundaries. However, this does NOT exempt bugs in the specific file(s) being tested - those must be fixed in the same PR. Only bugs in unrelated files may be deferred to separate issues.

**Comment by MonishPuttu** [comment]
> @MonishPuttu
> 
> The aim of testing is to validate the operation of the file under test and modify it if errors are found.
> 
> In this PR, please modify `src/graphql/types/Mutation/updateTag.ts` to fix the discovery that was also recorded in #4060

understood, will do

**Comment by MonishPuttu** [comment]
@palisadoes could you please, Add the 'ignore-sensitive-files-pr' label to this PR.
without it this cannot be fixed  `implementation bug in src/graphql/types/Mutation/updateTag.ts`

File: test/graphql/types/Mutation/updateTag.test.ts [hunk 1]
```diff
@@ -0,0 +1,946 @@
...
+			expect(result.errors).toBeUndefined();
+			expect(result.data?.updateTag).toEqual(
+				expect.objectContaining({
+					id: tagId,
+					folder: expect.objectContaining({
+						id: folderId,
+					}),
+				}),
+			);
+		});
```
[the remark this answers is not in the document]
Comment (MonishPuttu) [reply test/graphql/types/Mutation/updateTag.test.ts]: working on this

**Comment by palisadoes** [comment]
@coderabbitai full review using this prompt:
- https://github.com/PalisadoesFoundation/talawa-api/pull/4055#issuecomment-3686672073
